src/pgduckdb_types.cpp

```diff
@@ -173,24 +148,209 @@ ConvertNumeric(T value, idx_t scale) {
 		fractional_digits[i] = uint16_t(fractional_part % NBASE);
 		fractional_part /= NBASE;
 	}
-
 	ndigits = integral_ndigits + fractional_ndigits;
-
 	result.buf = (NumericDigit *)palloc(ndigits * sizeof(NumericDigit));
 	result.digits = result.buf;
 	auto &digits = result.digits;
-
 	idx_t digits_idx = 0;
 	for (idx_t i = integral_ndigits; i > 0; i--) {
 		digits[digits_idx++] = integral_digits[i - 1];
 	}
 	for (idx_t i = fractional_ndigits; i > 0; i--) {
 		digits[digits_idx++] = fractional_digits[i - 1];
 	}
-
 	return result;
 }
 
+static Datum
+ConvertNumericDatum(const duckdb::Value &value) {
+	if (value.type().id() == duckdb::LogicalTypeId::DOUBLE) {
+		return ConvertDoubleDatum(value);
+	}
+	NumericVar numeric_var;
+	D_ASSERT(value.type().id() == duckdb::LogicalTypeId::DECIMAL ||
+	         value.type().id() == duckdb::LogicalTypeId::HUGEINT ||
+	         value.type().id() == duckdb::LogicalTypeId::UBIGINT);
+	auto physical_type = value.type().InternalType();
+	const bool is_decimal = value.type().id() == duckdb::LogicalTypeId::DECIMAL;
+	uint8_t scale = is_decimal ? duckdb::DecimalType::GetScale(value.type()) : 0;
+
+	switch (physical_type) {
+	case duckdb::PhysicalType::INT16: {
+		numeric_var = ConvertNumeric<int16_t>(value.GetValueUnsafe<int16_t>(), scale);
+		break;
+	}
+	case duckdb::PhysicalType::INT32: {
+		numeric_var = ConvertNumeric<int32_t>(value.GetValueUnsafe<int32_t>(), scale);
+		break;
+	}
+	case duckdb::PhysicalType::INT64: {
+		numeric_var = ConvertNumeric<int64_t>(value.GetValueUnsafe<int64_t>(), scale);
+		break;
+	}
+	case duckdb::PhysicalType::UINT64: {
+		numeric_var = ConvertNumeric<uint64_t>(value.GetValueUnsafe<uint64_t>(), scale);
+		break;
+	}
+	case duckdb::PhysicalType::INT128: {
+		numeric_var = ConvertNumeric<hugeint_t, DecimalConversionHugeint>(value.GetValueUnsafe<hugeint_t>(), scale);
+		break;
+	}
+	default: {
+		throw duckdb::InvalidInputException("(PGDuckDB/ConvertDuckToPostgresValue) Unrecognized physical type for DECIMAL value");
+	}
+	}
+	auto numeric = CreateNumeric(numeric_var, NULL);
+	auto datum = NumericGetDatum(numeric);
+	return datum;
+}
+
+static Datum
+ConvertUUIDDatum(const duckdb::Value &value) {
+	D_ASSERT(value.type().id() == duckdb::LogicalTypeId::UUID);
+	D_ASSERT(value.type().InternalType() == duckdb::PhysicalType::INT128);
+	auto duckdb_uuid = value.GetValue<hugeint_t>();
+	pg_uuid_t *postgres_uuid = (pg_uuid_t *)palloc(sizeof(pg_uuid_t));
+
+	duckdb_uuid.upper ^= (uint64_t(1) << 63);
+	// Convert duckdb_uuid to bytes and store in postgres_uuid.data
+	uint8_t *uuid_bytes = (uint8_t *)&duckdb_uuid;
+
+	for (int i = 0; i < UUID_LEN; ++i) {
+		postgres_uuid->data[i] = uuid_bytes[UUID_LEN - 1 - i];
+	}
+
+	return UUIDPGetDatum(postgres_uuid);
+}
+
+
+struct PGTypeInfo {
+	int16 typlen;
+	bool typbyval;
+	char typalign;
+};
+
+PGTypeInfo
+GetPGTypeInfo(Oid typid) {
+	PGTypeInfo info;
+	HeapTuple tp;
+	Form_pg_type typtup;
+
+	tp = SearchSysCache1(TYPEOID, ObjectIdGetDatum(typid));
```

Comment:
This is one thing we shouldn't be doing. Because to safely call this function we need to take a global lock, which is way to slow in the hot codepath where you're currently using it. There are two options I think:

1. Make sure we call this `SearchSysCache1` function only once (while taking the lock) and cache its result somewhere ourselves which we can access in a thread-safe way.
2. Hardcode the values for the supported types for now.

Reply:
Yes, I use the second way to solve this problem based on the performance.